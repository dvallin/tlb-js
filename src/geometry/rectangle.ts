import { AbstractShape } from './shape'
import { Vector } from '../spatial'
import { Direction } from '../spatial/direction'

export class Rectangle extends AbstractShape {
  public static fromBounds(left: number, right: number, top: number, bottom: number): Rectangle {
    return new Rectangle(left, top, right - left + 1, bottom - top + 1)
  }

  public static footprint(position: Vector, direction: Direction, size: Vector): Rectangle {
    const up: Vector = Vector.fromDirection(direction).abs()
    const right: Vector = Vector.fromDirection(direction)
      .perpendicular()
      .abs()
    const cx = Math.floor(size.x / 2)
    const cy = Math.floor(size.y / 2)
    const topLeft = position.add(right.mult(-cx)).add(up.mult(-cy))
    const bottomRight = topLeft.add(right.mult(size.x - 1)).add(up.mult(size.y - 1))
    return Rectangle.fromBounds(topLeft.x, bottomRight.x, topLeft.y, bottomRight.y)
  }

  public static centerAt(x: number, y: number, size: number): Rectangle {
    return new Rectangle(x - size, y - size, 2 * size + 1, 2 * size + 1)
  }

  public constructor(public readonly x: number, public readonly y: number, public readonly width: number, public readonly height: number) {
    super()
  }

  public get left(): number {
    return this.x
  }

  public get right(): number {
    return this.x + this.width - 1
  }

  public get top(): number {
    return this.y
  }

  public get bottom(): number {
    return this.y + this.height - 1
  }

  public get topLeft(): Vector {
    return new Vector([this.left, this.top])
  }

  public get topRight(): Vector {
    return new Vector([this.right, this.top])
  }

  public get bottomRight(): Vector {
    return new Vector([this.right, this.bottom])
  }

  public get bottomLeft(): Vector {
    return new Vector([this.left, this.bottom])
  }

  public get center(): Vector {
    const cx = (this.left + this.right) / 2
    const cy = (this.top + this.bottom) / 2
    return new Vector([Math.floor(cx), Math.floor(cy)])
  }

  public centerOf(direction: Direction): Vector {
    switch (direction) {
      case 'up':
        return this.centerTop
      case 'right':
        return this.centerRight
      case 'down':
        return this.centerBottom
      case 'left':
        return this.centerLeft
    }
  }

  public get centerLeft(): Vector {
    const cy = (this.top + this.bottom) / 2
    return new Vector([this.left, Math.floor(cy)])
  }

  public get centerRight(): Vector {
    const cy = (this.top + this.bottom) / 2
    return new Vector([this.right, Math.floor(cy)])
  }

  public get centerTop(): Vector {
    const cx = (this.left + this.right) / 2
    return new Vector([Math.floor(cx), this.top])
  }

  public get centerBottom(): Vector {
    const cx = (this.left + this.right) / 2
    return new Vector([Math.floor(cx), this.bottom])
  }

  public plus(other: Rectangle): Rectangle {
    return Rectangle.fromBounds(
      Math.min(this.left, other.left),
      Math.max(this.right, other.right),
      Math.min(this.top, other.top),
      Math.max(this.bottom, other.bottom)
    )
  }

  public intersect(other: Rectangle): Rectangle {
    return Rectangle.fromBounds(
      Math.max(this.left, other.left),
      Math.min(this.right, other.right),
      Math.max(this.top, other.top),
      Math.min(this.bottom, other.bottom)
    )
  }

  public cover(other: Vector): Rectangle {
    return Rectangle.fromBounds(
      Math.min(this.left, other.x),
      Math.max(this.right, other.x),
      Math.min(this.top, other.y),
      Math.max(this.bottom, other.y)
    )
  }

  public bounds(): Rectangle {
    return this
  }

  public containsVector(p: Vector): boolean {
    return p.x >= this.x && p.x < this.x + this.width && p.y >= this.y && p.y < this.y + this.height
  }

  public translate(t: Vector): Rectangle {
    return new Rectangle(this.x + t.x, this.y + t.y, this.width, this.height)
  }

  public grow(cells: number = 1): Rectangle {
    return new Rectangle(this.x - cells, this.y - cells, this.width + 2 * cells, this.height + 2 * cells)
  }

  public shrink(cells: number = 1): Rectangle {
    const w = this.width - 2 * cells
    const h = this.height - 2 * cells
    return new Rectangle(w > 0 ? this.x + cells : this.x, h > 0 ? this.y + cells : this.y, w > 0 ? w : 0, h > 0 ? h : 0)
  }

  public all(f: (position: Vector) => boolean): boolean {
    for (let i = 0; i < this.height; i++) {
      for (let j = 0; j < this.width; j++) {
        const position = new Vector([j + this.x, i + this.y])
        if (!f(position)) {
          return false
        }
      }
    }
    return true
  }
}
