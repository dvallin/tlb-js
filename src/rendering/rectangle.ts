import { Rectangle } from "@/geometry/Rectangle"
import { Position } from "@/geometry/Position"
import { RenderIterator } from "@/rendering"

export function rasterize(rectangle: Rectangle, fill: boolean = false): RenderIterator {
    let y = rectangle.top
    let x = rectangle.left
    let done = false

    return () => {
        if (done) {
            return undefined
        }
        const result = new Position(Math.floor(x), Math.floor(y))
        x++
        if (x > rectangle.right) {
            x = rectangle.left
            y++
        }
        if (!fill && x > rectangle.left && y !== rectangle.top && y !== rectangle.bottom) {
            x = rectangle.right
        }
        if (y > rectangle.bottom) {
            done = true
        }
        return result
    }
}
