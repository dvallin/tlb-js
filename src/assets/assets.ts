import { Asset, Cover } from '../components/asset'
import { Vector } from '../spatial'
import { features, generators } from './features'
import { Feature } from '../components/feature'

const s1x1 = new Vector([1, 1])
const s1x2 = new Vector([1, 1])
const s3x1 = new Vector([3, 1])
const s2x2 = new Vector([2, 2])
function asset(name: string, size: Vector, cover: Cover, feature: (index: number) => Feature): Asset {
  return { name, size, cover, hasInventory: true, feature }
}

const assetsDefinition = {
  door: asset('metal door', s3x1, 'full', () => features['door']),
  locker: asset('a locker', s1x1, 'full', () => features['locker']),
  trash: asset('some trash', s1x1, 'none', () => features['trash']),
  loot: asset('a dead body', s1x1, 'none', () => features['loot']),
  table: asset('a table', s1x2, 'partial', () => features['table']),
  generator: asset('a generator', s2x2, 'full', generators['block']),
}

export type AssetType = keyof typeof assetsDefinition
export const assets: { [key in AssetType]: Asset } = assetsDefinition
