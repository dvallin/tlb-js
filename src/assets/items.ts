import { Effect } from '../components/effects'
import { ActionType } from './actions'
import { Item, WearableKind } from '../components/items'

function weapon(name: string, description: string, weight: number, effects: Effect[], actions: ActionType[], attachments: number): Item {
  return { kind: 'weapon', name, description, actions, effects, weight, attachments }
}

function consumable(name: string, description: string, weight: number, effects: Effect[]): Item {
  return { kind: 'consumable', attachments: 0, actions: ['consume'], name, description, effects, weight }
}

function equipment(
  kind: WearableKind,
  name: string,
  description: string,
  weight: number,
  effects: Effect[],
  actions: ActionType[],
  attachments: number
): Item {
  return { kind, name, description, actions, effects, weight, attachments }
}

export const itemsDefinition = {
  nailGun: weapon('nail gun', 'a tool to drive nails into wood, control panels or bodies.', 5, [], ['overcharge', 'hitAndRun', 'bolt'], 1),
  rifle: weapon('rifle', 'a regular rifle.', 5, [], ['shoot'], 2),
  sniperRifle: weapon('sniper rifle', '', 6, [], ['headshot', 'shoot'], 2),
  deathPill: consumable('death pill', 'a pill that kills you', 1, [{ type: 'kill', global: true, negated: false }]),
  bandages: consumable('bandages', 'a bandage to stop bleeding', 1, [{ type: 'bleed', global: false, negated: true }]),
  leatherJacket: equipment(
    'armor',
    'leather jacket',
    'a nice leather jacket',
    3,
    [{ type: 'defend', global: false, negated: false, value: 1 }],
    ['tighten'],
    1
  ),
  bootsOfStriding: equipment(
    'boots',
    'boots of striding',
    'Original Boots Of Striding TM, authentic D&D merch but not at all mint anymore',
    3,
    [],
    ['strideAndSlip'],
    2
  ),
}
export type ItemType = keyof typeof itemsDefinition
export const items: { [key in ItemType]: Item } = itemsDefinition
