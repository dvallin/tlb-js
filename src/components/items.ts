import { Entity } from '../ecs/entity'
import { Effect } from './effects'
import { ItemType } from '../assets/items'
import { ActionType } from '../assets/actions'

export type WearableKind = 'glove' | 'helmet' | 'armor' | 'pants' | 'boots'
export type ItemKind = 'scrap' | 'consumable' | 'weapon' | WearableKind

export interface Item {
  kind: ItemKind
  name: string
  description: string
  actions: ActionType[]
  effects: Effect[]
  weight: number
  attachments: number
}

export interface InventoryComponent {
  content: Entity[]
}

export interface ItemComponent {
  type: ItemType
}

export interface EquipmentAttachement {
  entity: Entity

  bodyParts: string[]
}

export interface EquipedItemsComponent {
  equipment: EquipmentAttachement[]
}
