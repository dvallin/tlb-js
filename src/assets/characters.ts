import { BodyPart, CharacterStats, CharacterStatsComponent, createCharacterStatsComponent } from '../components/character-stats'
import { TlbWorld } from '../tlb'
import { features, FeatureType } from './features'
import { FeatureComponent } from '../components/feature'
import { FovComponent } from '../components/fov'
import { HasActionComponent } from '../components/action'
import { ItemComponent, InventoryComponent, EquipedItemsComponent } from '../components/items'
import { ActiveEffectsComponent } from '../components/effects'
import { CharacterCreator } from '../generative/complex-embedder'
import { Entity } from '../ecs/entity'
import { AiComponent } from '../components/ai'
import { ActionType } from './actions'
import { ItemType } from './items'
import { addDialog } from './dialogs'
import { TriggersComponent } from '../components/trigger'

function humanoidBodyParts(core: number, head: number, limp: number): { [key: string]: BodyPart } {
  return {
    head: { type: 'head', itemAttachments: ['helmet'], health: head },
    leftArm: { type: 'arm', itemAttachments: ['glove', 'weapon'], health: limp },
    rightArm: { type: 'arm', itemAttachments: ['glove', 'weapon'], health: limp },
    torso: { type: 'torso', itemAttachments: ['armor'], health: core },
    leftLeg: { type: 'leg', itemAttachments: ['pants', 'boots'], health: limp },
    rightLeg: { type: 'leg', itemAttachments: ['pants', 'boots'], health: limp },
  }
}

const charactersStatsDefinition = {
  player: {
    bodyParts: humanoidBodyParts(14, 5, 5),
    strength: 5,
    movement: 4,
    actions: 4,
    aim: 10,
  },
  guard: {
    bodyParts: humanoidBodyParts(3, 1, 1),
    strength: 5,
    movement: 3,
    actions: 3,
    aim: 6,
  },
  eliteGuard: {
    bodyParts: humanoidBodyParts(5, 3, 3),
    strength: 8,
    movement: 2,
    actions: 5,
    aim: 6,
  },
  boss: {
    bodyParts: humanoidBodyParts(8, 5, 5),
    strength: 10,
    movement: 3,
    actions: 9,
    aim: 8,
  },
}
export type CharacterStatsType = keyof typeof charactersStatsDefinition
export const characterStats: { [key in CharacterStatsType]: CharacterStats } = charactersStatsDefinition

const characterCreatorsDefinition = {
  player: (world: TlbWorld) => createPlayer(world),
  eliteGuard: (world: TlbWorld) => {
    const entity = createDefaultNpc(world, 'some elite guard', 'eliteGuard', 'eliteGuard')
    addDialog(world, entity, 'guardRandomRemarks')
    return entity
  },
  guard: (world: TlbWorld) => {
    const entity = createDefaultNpc(world, 'some guard', 'guard', 'guard')
    addDialog(world, entity, 'guardRandomRemarks')
    return entity
  },
}

export const defaultActions: ActionType[] = ['longMove', 'hit', 'rush', 'endTurn']

export function createPlayer(world: TlbWorld): Entity {
  const player = createCharacter(world, 'player', 'player', defaultActions)
  world
    .editEntity(player)
    .withComponent<{}>('player', {})
    .withComponent<FovComponent>('fov', { fov: [] })
    .withComponent<{}>('viewport-focus', {})

  take(world, player, 'idCard')
  take(world, player, 'rifle')
  take(world, player, 'sniperRifle')
  take(world, player, 'bootsOfStriding')
  equip(world, player, 'nailGun', ['leftArm'])
  equip(world, player, 'leatherJacket', ['torso', 'leftArm', 'rightArm'])
  return player
}

export function createDefaultNpc(world: TlbWorld, name: string, statsType: CharacterStatsType, featureType: FeatureType): Entity {
  const npc = createEmptyNpc(world, name, statsType, featureType, defaultActions)
  equip(world, npc, 'rifle', ['leftArm'])
  return npc
}

export function take(world: TlbWorld, entity: Entity, type: ItemType): Entity {
  const item = world.createEntity().withComponent<ItemComponent>('item', { type }).entity
  world.getComponent<InventoryComponent>(entity, 'inventory')!.content.push(item)
  return item
}

export function equip(world: TlbWorld, entity: Entity, type: ItemType, bodyParts: string[]): void {
  const item = take(world, entity, type)
  world.getComponent<EquipedItemsComponent>(entity, 'equiped-items')!.equipment.push({
    entity: item,
    bodyParts,
  })
}

export function createEmptyNpc(
  world: TlbWorld,
  name: string,
  statsType: CharacterStatsType,
  featureType: FeatureType,
  actions: ActionType[]
): Entity {
  const character = createCharacter(world, statsType, featureType, actions)
  return world
    .editEntity(character)
    .withComponent('npc', {})
    .withComponent<TriggersComponent>('triggers', { entities: [], type: 'dialog', name })
    .withComponent<AiComponent>('ai', { type: 'rushing', state: 'idle', interest: undefined, distrust: 0 }).entity
}

export function createCharacter(world: TlbWorld, statsType: CharacterStatsType, featureType: FeatureType, actions: ActionType[]): Entity {
  return world
    .createEntity()
    .withComponent<HasActionComponent>('has-action', { actions })
    .withComponent<InventoryComponent>('inventory', { content: [] })
    .withComponent<EquipedItemsComponent>('equiped-items', { equipment: [] })
    .withComponent<FeatureComponent>('feature', { feature: () => features[featureType] })
    .withComponent<CharacterStatsComponent>('character-stats', createCharacterStatsComponent(statsType))
    .withComponent<ActiveEffectsComponent>('active-effects', { effects: [] }).entity
}

export type CharacterCreatorType = keyof typeof characterCreatorsDefinition
export const characterCreators: { [key in CharacterCreatorType]: CharacterCreator } = characterCreatorsDefinition
