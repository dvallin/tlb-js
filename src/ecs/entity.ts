import { World } from "./world"

export type Entity = number

export class EntityModifier<C, S, R> {

    public constructor(
        private readonly world: World<C, S, R>,
        public readonly entity: Entity
    ) { }

    public withComponent<T>(name: C, component: T): EntityModifier<C, S, R> {
        this.world.getStorage(name).insert(this.entity, component)
        return this
    }

    public removeComponent(name: C): EntityModifier<C, S, R> {
        this.world.getStorage(name).remove(this.entity)
        return this
    }

    public delete(): void {
        this.world.components.forEach(s => s.remove(this.entity))
    }
}
