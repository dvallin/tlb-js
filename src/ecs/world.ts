import { Storage } from './storage'
import { System } from './system'
import { Resource } from './resource'
import { Entity, EntityModifier } from './entity'

export class World<C, S, R> {
  public readonly components: Map<C, Storage<object>> = new Map()
  public readonly resources: Map<R, Resource<C, S, R>> = new Map()
  public readonly systems: Map<S, System<C, S, R>> = new Map()
  public readonly emptySystems: Set<S> = new Set()
  public readonly activeSystems: Set<S> = new Set()

  private openEntities: Set<Entity> = new Set()
  private lastEntity: Entity = -1

  public get entityCount(): number {
    return this.lastEntity + 1 - this.openEntities.size
  }

  public registerComponentStorage<T extends object>(component: C, storage: Storage<T>): void {
    this.components.set(component, storage)
  }

  public getStorage<T extends object>(component: C): Storage<T> {
    return this.components.get(component)! as Storage<T>
  }

  public getResource<T extends Resource<C, S, R>>(resource: R): T {
    return this.resources.get(resource)! as T
  }

  public getComponent<T extends object>(entity: Entity, component: C): T | undefined {
    const storage = this.getStorage<T>(component)
    if (storage !== undefined) {
      return storage.get(entity)
    }
    return undefined
  }

  public hasEntity(entity: Entity): boolean {
    return entity <= this.lastEntity && !this.openEntities.has(entity)
  }

  public hasComponent(entity: Entity, component: C): boolean {
    const storage = this.getStorage(component)
    if (storage !== undefined) {
      return storage.has(entity)
    }
    return false
  }

  public createEntity(): EntityModifier<C, S, R> {
    let entity
    if (this.openEntities.size > 0) {
      entity = this.openEntities.values().next().value
      this.openEntities.delete(entity)
    } else {
      entity = ++this.lastEntity
    }
    return new EntityModifier(this, entity)
  }

  public editEntity(entity: Entity): EntityModifier<C, S, R> {
    return new EntityModifier(this, entity)
  }

  public deleteEntity(entity: Entity): void {
    this.editEntity(entity).delete()
    this.openEntities.add(entity)
  }

  public registerSystem(name: S, system: System<C, S, R>): void {
    this.systems.set(name, system)
  }

  public enableSystem(name: S): void {
    this.activeSystems.add(name)
  }

  public disableSystem(name: S): void {
    this.activeSystems.delete(name)
  }

  public activeSystemsList(): S[] {
    const l: S[] = []
    this.activeSystems.forEach(s => l.push(s))
    return l
  }

  public registerResource(resource: Resource<C, S, R>): void {
    this.resources.set(resource.kind, resource)
  }

  public updateResources(): void {
    for (const resource of this.resources) {
      resource[1].update(this)
    }
  }

  public updateSystems(): void {
    this.emptySystems.clear()
    for (const name of this.activeSystems) {
      const system = this.systems.get(name)!
      const firstStorage = this.getStorage(system.components[0])
      let calls = 0
      firstStorage.foreach((entity: Entity) => {
        let callSystem = true
        for (let i = 1; i < system.components.length; i++) {
          if (!this.hasComponent(entity, system.components[i])) {
            callSystem = false
            break
          }
        }
        if (callSystem) {
          system.update(this, entity)
          calls++
        }
      })
      if (calls === 0) {
        this.emptySystems.add(name)
      }
    }
  }
}
