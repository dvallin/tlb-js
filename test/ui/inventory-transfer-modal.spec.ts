import { InventoryTransferModal } from '../../src/ui/inventory-transfer-modal'
import { Rectangle } from '../../src/geometry/rectangle'
import { characterCreators } from '../../src/assets/characters'
import { WorldMapResource } from '../../src/resources/world-map'
import { registerComponents, TlbWorld } from '../../src/tlb'
import { World } from '../../src/ecs/world'
import { stringRenderer } from '../render'
import { mockInput, mockImplementation, mockReturnValue } from '../mocks'
import { Input, KeyboardCommand, NumericKeyboardCommand } from '../../src/resources/input'
import { InventoryComponent } from '../../src/components/items'
import { Entity } from '../../src/ecs/entity'

describe('InventoryTransferModal', () => {
  let world: TlbWorld
  let modal: InventoryTransferModal
  let input: Input
  let guard: Entity
  beforeEach(() => {
    world = new World()
    registerComponents(world)

    world.registerResource(new WorldMapResource(4))
    input = mockInput(world)

    const player = characterCreators.player(world)
    guard = characterCreators.guard(world)
    modal = InventoryTransferModal.build(new Rectangle(0, 0, 30, 10), player, 'player', guard, 'guard')
  })

  it('displays two inventories', () => {
    modal.update(world)

    const { renderer, getDisplay } = stringRenderer()
    modal.render(renderer)
    expect(getDisplay()).toMatchSnapshot()
  })

  it('highlights right inventory', () => {
    mockImplementation<KeyboardCommand, boolean>(input.isActive, key => key === 'right')

    modal.update(world)

    const { renderer, getDisplay } = stringRenderer()
    modal.render(renderer)
    expect(getDisplay()).toMatchSnapshot()
  })

  it('transfers items', () => {
    mockReturnValue<NumericKeyboardCommand>(input.numericActive, 0)

    modal.update(world)

    expect(world.getComponent<InventoryComponent>(guard, 'inventory')!.content).toHaveLength(2)
  })

  it('does not transfer equiped items', () => {
    mockReturnValue<NumericKeyboardCommand>(input.numericActive, 4)

    modal.update(world)

    expect(world.getComponent<InventoryComponent>(guard, 'inventory')!.content).toHaveLength(1)
  })
})
