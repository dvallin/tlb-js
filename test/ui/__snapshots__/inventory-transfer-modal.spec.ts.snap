// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InventoryTransferModal displays two inventories 1`] = `
"      /player/      /guard/
 1 id card (0) rifle (5)
 2 rifle (5)        
 3 sniper rifle (6) 
 4 boots of striding (3)
 5 nail gun (5)     
 6 leather jacket (3)


        22/25       5/25"
`;

exports[`InventoryTransferModal highlights right inventory 1`] = `
"      /player/      /guard/
 id card (0)   1 rifle (5)
 rifle (5)          
 sniper rifle (6)   
 boots of striding (3)
 nail gun (5)       
 leather jacket (3) 


        22/25       5/25"
`;
